builder/cloudstack/step_create_instance.go

```diff
@@ -2,16 +2,66 @@ package cloudstack
 
 import (
 	"encoding/base64"
+	"errors"
 	"fmt"
+	"net"
 	"strings"
 
+	"github.com/hashicorp/packer/common"
 	"github.com/hashicorp/packer/packer"
+	"github.com/hashicorp/packer/template/interpolate"
 	"github.com/mitchellh/multistep"
 	"github.com/xanzy/go-cloudstack/cloudstack"
 )
 
+// createInstanceTemplateData represents variables for user_data interpolation
+type createInstanceTemplateData struct {
+	HTTPIP   string
+	HTTPPort uint
+}
+
 // stepCreateInstance represents a Packer build step that creates CloudStack instances.
-type stepCreateInstance struct{}
+type stepCreateInstance struct {
+	Ctx interpolate.Context
+}
+
+func hostIP() (string, error) {
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return "", err
+	}
+	// loop through all interface and search for non-loopback ipv4 addresses
+	for _, iface := range interfaces {
+		if iface.Flags&net.FlagUp == 0 {
+			continue
+		}
+		if iface.Flags&net.FlagLoopback != 0 {
+			continue
+		}
+		addresses, err := iface.Addrs()
+		if err != nil {
+			return "", err
+		}
+		for _, address := range addresses {
+			var ip net.IP
+			switch v := address.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+			if ip == nil || ip.IsLoopback() {
+				continue
+			}
+			ip = ip.To4()
+			if ip == nil {
+				continue
+			}
+			return ip.String(), nil
+		}
+	}
+	return "", errors.New("IP not found")
+}
```

Comment:
May I suggest a small refactored func for this one?

```go
func hostIP() (string, error) {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		return "", err
	}

	for _, addr := range addrs {
		if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
			if ipnet.IP.To4() != nil {
				return ipnet.IP.String(), nil
			}
		}
	}

	return "", errors.New("No host IP found")
}
```

This should do the exact same thing, but is a bit cleaner and more readable (IMO). Also could you move this helper function to the end of this file? Thx!